### PR TITLE
logging: wire klog backend, but only output request throttling logs

### PR DESCRIPTION
**Pull request by sttts**
### Description of your changes

Vendoring and wiring https://github.com/crossplane/crossplane-runtime/pull/673.

I have: 

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- ~~[ ] Added or updated unit tests.~~
- ~~[ ] Added or updated e2e tests.~~
- ~~[ ] Linked a PR or a [docs tracking issue] to [document this change].~~
- ~~[ ] Added `backport release-x.y` labels to auto-backport this PR.~~

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/master/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/master/contributing#checklist-cheat-sheet

### CODE REVIEW

**Review by jbw976 — COMMENTED** [review]
@sttts would it be possible for you to provide more details in the PR description that show what the logs look like with this change? besides helping understand the general experience, I'm also hoping to see how these logs will be able to positively identify slowdowns and if there's anything the person reading the logs can do about it.

**Comment by sttts** [comment]
@phisco added the logging lines from client-go that now pass.

**Comment by sttts** [comment]
As an example, @turkenh would have seen something like
```
Waited for 1.033772408s due to client-side throttling, not priority and fairness, request: GET:https://api.example.org/apis/pkg.crossplane.io/v1?timeout=32s
```
in https://github.com/crossplane-contrib/provider-kubernetes/pull/203 like crazy.

**Comment by turkenh** [comment]
> As an example, @turkenh would have seen something like

@sttts I am not sure, my understanding was we were [rate-limiting](As an example, @turkenh would have seen something like) ourself even before making client side requests 🤔

**Comment by sttts** [comment]
Ok, controller rate limiting is even before, yes.

**Comment by sttts** [comment]
For background: there are 3 levels of throttling usually:

1. Controller rate limiting (max-reconcile-rate)
2. Client-side request throttling
3. Server-side request throttling. 

In Crossplane afaik request rps is 5x Max-reconcile-rate. Any of these 3 can make things slow if load is high enough.

File: cmd/crossplane/main.go [hunk 1]
```diff
@@ -68,6 +73,11 @@ func (d debugFlag) BeforeApply(ctx *kong.Context) error { //nolint:unparam // Be
 	// *very* verbose even at info level, so we only provide it a real
 	// logger when we're running in debug mode.
 	ctrl.SetLogger(zl)
+
```
Comment (negz) [inline cmd/crossplane/main.go:76]: A comment explaining what this is doing and why would be helpful.

File: cmd/crossplane/main.go [hunk 2]
```diff
@@ -84,6 +94,13 @@ func main() {
 	// logger is *very* verbose even at info level. This is not really needed,
 	// but otherwise we get a warning from the controller-runtime.
 	ctrl.SetLogger(zap.New(zap.WriteTo(io.Discard)))
+
+	fs := flag.NewFlagSet(os.Args[0], flag.ExitOnError)
+	klog.InitFlags(fs)
+	fs.Parse([]string{"--v=3"}) // nolint:errcheck
+	klogr := logr.New(&logrFilter{zl.GetSink()})
+	klog.SetLogger(klogr)
```
Comment (negz) [inline cmd/crossplane/main.go:102]: Ditto here WRT adding an explanatory comment.

It looks a bit like we're discarding `v=4` and lower, and sending `v=3` and lower to our zap logger? Is that right?

File: cmd/crossplane/main.go [hunk 3]
```diff
@@ -109,3 +126,35 @@ func main() {
 	ctx.FatalIfErrorf(apis.AddToScheme(s), "cannot add Crossplane API types to scheme")
 	ctx.FatalIfErrorf(ctx.Run(s))
 }
+
+type logrFilter struct {
```
Comment (negz) [inline cmd/crossplane/main.go:130]: A comment here mentioning that this filter is specifically filtering down to only request throttling would be helpful. Possibly just a more explicit type name would be enough too (`throttlingLogFilter` or something).

**Review by negz — COMMENTED** [review]
This seems reasonable. Do you want to go straight to adding it to crossplane-runtime pkg/logging rather than proceeding here?

If you want to proceed here, see comments. There's also some linter errors to address.

WRT the log output -  it will use the same format as the rest of our log lines, right? e.g. If we set our logger to use structured logging these would be structured too? I expect that to be the case from the code but it's not clear from the log lines in the description.

**Comment by negz** [comment]
> In Crossplane afaik request rps is 5x Max-reconcile-rate. Any of these 3 can make things slow if load is high enough.

FYI I retroactively documented our client-side rate limiters in https://github.com/crossplane/crossplane/pull/5415.

I was thinking debug logging and emitting metrics when the workqueue and/or rate-limiting `Reconciler` "middleware" were rate limiting would be useful too.

**Comment by sttts** [comment]
Would go directly to crossplane-runtime. Will open it there.

File: cmd/crossplane/main.go (hunk 2 above)
Comment (sttts) [reply cmd/crossplane/main.go:102]: Yes.

Comment (sttts) [reply cmd/crossplane/main.go:102]: done

File: cmd/crossplane/main.go (hunk 3 above)
Comment (sttts) [reply cmd/crossplane/main.go:130]: done

**Comment by sttts** [comment]
@turkenh @negz update to use the crossplane-runtime helper.

**Review by negz — APPROVED** [review]
Nice, thank you!

**Pull request merged**